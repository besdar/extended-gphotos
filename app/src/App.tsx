import React, { FC } from 'react';
import { getRandomKinopoiskImageInfo } from './api/kinopoisk-films-data';
import { FilmsData } from './api/kinopoisk-films-object';
import ImageGallery from './Components/ImageGallery/ImageGallery';

const App: FC = () => (
  <ImageGallery
    getRandomImage={getRandomKinopoiskImageInfo(FilmsData)}
    galleryStructureConfig={[
      window.innerHeight,
      window.innerWidth,
      window.innerWidth / 3 > 300 ? 100 : 50,
      3,
    ]}
  />
);

export default App;
