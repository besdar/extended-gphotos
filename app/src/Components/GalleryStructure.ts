const isSquareFit = (squaresMap: number[][], currentRow: number, currentColumn: number, elementarySquareSideCount: number) => {
    for (let i = currentRow; i < currentRow + elementarySquareSideCount; i++) {
        for (let j = currentColumn; j < currentColumn + elementarySquareSideCount; j++) {
            try {
                if (squaresMap[i][j] !== 0) {
                    return false;
                }
            } catch {
                return false;
            }
        }
    }

    return true;
};

const getRandomSquareSide = (minSquareSide: number, timesNumberMaxSideGreaterThanMinSide: number): number => {
    if (timesNumberMaxSideGreaterThanMinSide === 0) {
        return minSquareSide;
    }

    return minSquareSide * Math.round(Math.random()) + getRandomSquareSide(minSquareSide, timesNumberMaxSideGreaterThanMinSide - 1);
};

const generateSquaresMap = (mapHeight: number, mapWidth: number, squareSideLength: number) => {
    const rowLength = Math.floor(mapWidth / squareSideLength);

    if (rowLength === 0) {
        return [];
    }

    const rowsCount = Math.floor(mapHeight / squareSideLength);
    const squaresMap = [] as number[][];

    for (let i = 0; i < rowsCount; i++) {
        squaresMap.push(new Array(rowLength).fill(0));
    }

    return squaresMap;
};

export type SquareType = {
    xstart: number,
    xend: number,
    ystart: number,
    yend: number,
}

export const getGalleryStructureData = (mapHeight: number, mapWidth: number, minSquareSide: number, timesNumberMaxSideGreaterThanMinSide: number) => {
    const squaresMap = generateSquaresMap(mapHeight, mapWidth, minSquareSide);
    const result = {
        imagesStructureData: [] as SquareType[],
        rowLength: 0,
    }

    if (squaresMap.length === 0) {
        return result;
    }

    result.rowLength = squaresMap[0].length;
    const rowsCount = squaresMap.length;
    let currentRow = 0;
    let currentColumn = 0;
    let mapNumber = 1;

    while (true) {
        const squareSide = getRandomSquareSide(minSquareSide, timesNumberMaxSideGreaterThanMinSide);
        const sideInElementarySquares = squareSide / minSquareSide;

        if (isSquareFit(squaresMap, currentRow, currentColumn, sideInElementarySquares)) {
            for (let j = 0; j < sideInElementarySquares; j++) {
                for (let i = 0; i < sideInElementarySquares; i++) {
                    squaresMap[currentRow + i][currentColumn + j] = mapNumber
                }
            }

            result.imagesStructureData.push({
                xstart: currentColumn + 1,
                xend: currentColumn + 1 + sideInElementarySquares,
                ystart: currentRow + 1,
                yend: currentRow + 1 + sideInElementarySquares,
            });
            currentColumn = squaresMap[currentRow].findIndex(el => el === 0);
            mapNumber = mapNumber + 1;
        }

        if ((currentColumn === result.rowLength || currentColumn === -1) && currentRow === rowsCount - 1) {
            break;
        }

        if (currentColumn === result.rowLength || currentColumn === -1) {
            currentRow = currentRow + 1;
            currentColumn = squaresMap[currentRow].findIndex(el => el === 0);
        }
    }

    return result;
};