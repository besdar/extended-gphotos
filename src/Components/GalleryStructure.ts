const isSquareFit = (
  squaresMap: number[][],
  currentRow: number,
  currentColumn: number,
  elementarySquareSideCount: number,
) => {
  for (let i = currentRow; i < currentRow + elementarySquareSideCount; i += 1) {
    for (let j = currentColumn; j < currentColumn + elementarySquareSideCount; j += 1) {
      try {
        if (squaresMap[i][j] !== 0) {
          return false;
        }
      } catch {
        return false;
      }
    }
  }

  return true;
};

const getRandomSquareSide = (timesNumberMaxSideGreaterThanMinSide: number): number => {
  if (timesNumberMaxSideGreaterThanMinSide === 0) {
    return 1;
  }

  return Math.round(Math.random()) + getRandomSquareSide(timesNumberMaxSideGreaterThanMinSide - 1);
};

export const calculateGalleryStructureParams = (
  height: number,
  width: number,
  minSquareSide: number,
  maxSqureSide: number,
): galleryStructureParamsType | null => {
  if (minSquareSide <= 0) {
    return null;
  }

  const rowLength = Math.floor(width / minSquareSide);
  const maxSquareSizeInColumns = Math.floor(maxSqureSide / minSquareSide);

  if (rowLength <= 0 || maxSquareSizeInColumns <= 1) {
    return null;
  }

  const rowsCount = Math.floor(height / minSquareSide);

  return {
    columnsCount: rowLength,
    rowsCount,
    maxSquareSizeInColumns,
  };
};

const generateSquaresMap = (
  rowsCount: number,
  columnsCount: number,
) => {
  const squaresMap = [] as number[][];

  for (let i = 0; i < rowsCount; i += 1) {
    squaresMap.push(new Array(columnsCount).fill(0));
  }

  return squaresMap;
};

export type SquareType = {
  xstart: number,
  xend: number,
  ystart: number,
  yend: number,
}

export type galleryStructureParamsType = {
  rowsCount: number,
  columnsCount: number,
  maxSquareSizeInColumns: number,
};
export const getGalleryStructureData = ({
  rowsCount,
  columnsCount,
  maxSquareSizeInColumns,
}: galleryStructureParamsType) => {
  const result = [] as SquareType[];

  if (rowsCount <= 0) {
    return result;
  }

  const squaresMap = generateSquaresMap(rowsCount, columnsCount);
  let currentRow = 0;
  let currentColumn = 0;
  let mapNumber = 1;

  while ((currentColumn !== columnsCount && currentColumn !== -1) || currentRow !== rowsCount - 1) {
    const sideInElementarySquares = getRandomSquareSide(maxSquareSizeInColumns);
    const isEndOfLine = currentColumn === columnsCount || currentColumn === -1;

    if (isEndOfLine) {
      currentRow += 1;
      currentColumn = squaresMap[currentRow].findIndex((el) => el === 0);
    }

    if (isSquareFit(squaresMap, currentRow, currentColumn, sideInElementarySquares)) {
      for (let j = 0; j < sideInElementarySquares; j += 1) {
        for (let i = 0; i < sideInElementarySquares; i += 1) {
          squaresMap[currentRow + i][currentColumn + j] = mapNumber;
        }
      }

      result.push({
        xstart: currentColumn + 1,
        xend: currentColumn + 1 + sideInElementarySquares,
        ystart: currentRow + 1,
        yend: currentRow + 1 + sideInElementarySquares,
      });
      currentColumn = squaresMap[currentRow].findIndex((el) => el === 0);
      mapNumber += 1;
    }
  }

  return result;
};
