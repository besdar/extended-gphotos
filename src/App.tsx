import React, { FC } from 'react';
import { getRandomKinopoiskImageInfo } from './api/kinopoisk/kinopoisk-films-data';
import { FilmsData } from './api/kinopoisk/kinopoisk-films-object';
import { calculateGalleryStructureParams } from './Components/GalleryStructure';
import ImageGallery from './Components/ImageGallery/ImageGallery';

const App: FC = () => {
  const galleryStructureConfig = calculateGalleryStructureParams(
    window.innerHeight,
    window.innerWidth,
    window.innerWidth / 3 > 300 ? 100 : 50,
    window.innerWidth / 3 > 300 ? 300 : 150,
  );

  if (galleryStructureConfig === null) {
    return null;
  }

  return (
    <ImageGallery
      getRandomImage={getRandomKinopoiskImageInfo(FilmsData)}
      galleryStructureConfig={galleryStructureConfig}
      containerHeight={window.innerHeight}
    />
  );
};

export default App;
