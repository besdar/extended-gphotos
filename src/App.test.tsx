import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import { calculateGalleryStructureParams } from './Components/GalleryStructure';

jest.mock('./Components/GalleryStructure', () => ({
  calculateGalleryStructureParams: jest.fn()
}));

const calculateGalleryStructureParamsMock = calculateGalleryStructureParams as jest.MockedFunction<typeof calculateGalleryStructureParams>;

describe('Check App render with various helper function returns', () => {
  beforeEach(() => {
    calculateGalleryStructureParamsMock.mockReset();
  });
  
  it('App do not render with wrong gallery structure function call', () => {
    calculateGalleryStructureParamsMock.mockReturnValueOnce(null);
    const component = shallow(<App />);
  
    expect(component.isEmptyRender()).toBeTruthy();
  
    jest.unmock('./Components/GalleryStructure');
  });
  
  it('App renders with normal gallery structure function call', () => {
    calculateGalleryStructureParamsMock.mockReturnValueOnce({
      columnsCount: 2,
      maxSquareSizeInColumns: 2,
      rowsCount: 2,
    });
  
    const component = shallow(<App />);
  
    expect(component).toMatchSnapshot();
  });
});

describe('Tests with window resize', () => {
  const originalWindowInnerWidth = window.innerWidth;

  beforeAll(() => {
    calculateGalleryStructureParamsMock.mockRestore();
  });

  afterEach(() => {
    Object.defineProperty(window, 'innerWidth', {
      value: originalWindowInnerWidth,
    });
  });

  it('Small screen case', () => {
    calculateGalleryStructureParamsMock.mockRestore();
    Object.defineProperty(window, 'innerWidth', {
      value: 599,
    });

    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });

  it('Big screen case', () => {
    calculateGalleryStructureParamsMock.mockRestore();
    Object.defineProperty(window, 'innerWidth', {
      value: 601,
    });

    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });
})
