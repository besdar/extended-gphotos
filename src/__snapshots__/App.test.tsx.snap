// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check App render with various helper function returns App renders with normal gallery structure function call 1`] = `
<ImageGallery
  containerHeight={768}
  galleryStructureConfig={
    Object {
      "columnsCount": 2,
      "maxSquareSizeInColumns": 2,
      "rowsCount": 2,
    }
  }
  getRandomImage={[Function]}
/>
`;

exports[`Tests with window resize Big screen case 1`] = `
<ImageGallery
  containerHeight={768}
  getRandomImage={[Function]}
/>
`;

exports[`Tests with window resize Small screen case 1`] = `
<ImageGallery
  containerHeight={768}
  getRandomImage={[Function]}
/>
`;
